FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -0,0 +1,251 @@
+    request.setValue(token, forHTTPHeaderField: ModelInfoRetriever.fisTokenHTTPHeader)
+    if let info = modelInfo, info.hash.count > 0 {
+      request.setValue(info.hash, forHTTPHeaderField: ModelInfoRetriever.hashMatchHTTPHeader)
+    }
+    return request
+  }
+
+  /// Get model info from server.
+  func downloadModelInfo(completion: @escaping (DownloadError?) -> Void) {
+    /// Get FIS token.
+    installations.authToken { [weak self] tokenResult, error in
+      guard let result = tokenResult
+      else { completion(.internalError(description: ModelInfoRetriever.tokenErrorDescription))
+        return
+      }
+      /// Get model info fetch URL with appropriate HTTP headers.
+      guard let request = self?.getModelInfoFetchURLRequest(token: result.authToken)
+      else {
+        completion(.internalError(description: ModelInfoRetriever.modelFetchURLErrorDescription))
+        return
+      }
+      /// Download model info.
+      let dataTask = URLSession.shared.dataTask(with: request) { [weak self]
+        data, response, error in
+        if let downloadError = error {
+          completion(.internalError(description: downloadError.localizedDescription))
+        } else {
+          guard let httpResponse = response as? HTTPURLResponse else {
+            completion(.internalError(description: ModelInfoRetriever
+                .invalidHTTPResponseErrorDescription))
+            return
+          }
+
+          guard httpResponse.statusCode == 200 || httpResponse.statusCode == 304 else {
+            // TODO: Improve http status code error handling
+            completion(.notFound)
+            return
+          }
+
+          /// Local model not modified.
+          if httpResponse.statusCode == 304 {
+            return
+          }
+
+          guard let modelHash = httpResponse
+            .allHeaderFields[ModelInfoRetriever.etagHTTPHeader] as? String else {
+            completion(.internalError(description: ModelInfoRetriever
+                .missingModelHashErrorDescription))
+            return
+          }
+
+          guard let data = data else {
+            completion(.internalError(description: ModelInfoRetriever
+                .invalidHTTPResponseErrorDescription))
+            return
+          }
+
+          self?.saveModelInfo(data: data, modelHash: modelHash)
+        }
+      }
+      dataTask.resume()
+    }
+  }
+
+  /// Save model info to user defaults.
+  func saveModelInfo(data: Data, modelHash: String) {
+    // TODO: Save model info to user defaults
+    modelInfo?.hash = modelHash
+  }
+}
+
+/// Named user defaults for FirebaseML.
+extension UserDefaults {
+  static var firebaseMLDefaults: UserDefaults {
+    let suiteName = "com.google.firebase.ml"
+    let defaults = UserDefaults(suiteName: suiteName)!
+    return defaults
+  }
+
+  /// For testing: returns a new cleared instance of user defaults.
+  static func getTestInstance() -> UserDefaults {
```

Comment:
I would try to avoid adding test code to production version. Can it be defined in tests instead?

Reply:
Done.